starfish/core/spots/DecodeSpots/util.py

```diff
@@ -0,0 +1,166 @@
+from collections import defaultdict
+from typing import Any, Dict, Hashable, List, Mapping, Sequence, Tuple
+
+import numpy as np
+import pandas as pd
+from sklearn.neighbors import NearestNeighbors
+
+from starfish.core.intensity_table.intensity_table import IntensityTable
+from starfish.core.types import Axes, Features, SpotFindingResults
+
+
+def _match_spots(
+    round_dataframes: Dict[int, pd.DataFrame], anchor_round: int
+) -> Tuple[pd.DataFrame, pd.DataFrame]:
+    """ For each spot in anchor round, find the closest spot within search_radius in all rounds.
+
+    Parameters
+    ----------
+    round_dataframes : Dict[int, pd.DataFrame]
+        Output from _merge_spots_by_round, contains mapping of image volumes from each round to
+        all the spots detected in them.
+    search_radius : int
+        The maximum (euclidean) distance in pixels for a spot to be considered matching in
+        a round subsequent to the anchor round.
+    anchor_round : int
+        The imaging round to seed the local search from.
+
+    Returns
+    -------
+    pd.DataFrame
+        Spots x rounds dataframe containing the distances to the nearest spot. np.nan if
+        no spot is detected within search radius
+    pd.DataFrame
+        Spots x rounds dataframe containing the indices of the nearest spot to the
+        corresponding round_dataframe. np.nan if no spot is detected within search radius.
+
+    """
+    reference_df = round_dataframes[anchor_round]
+    reference_coordinates = reference_df[[Axes.ZPLANE, Axes.Y, Axes.X]]
+
+    dist = pd.DataFrame(
+        data=np.zeros((reference_df.shape[0], len(round_dataframes)), dtype=float),
+        columns=list(round_dataframes.keys())
+    )
+    ind = pd.DataFrame(
+        data=np.zeros((reference_df.shape[0], len(round_dataframes)), dtype=np.int32),
+        columns=list(round_dataframes.keys())
+    )
+
+    # fill data for anchor round; every spot is a perfect match to itself.
+    ind[anchor_round] = np.arange(reference_df.shape[0], dtype=np.int32)
+
+    # get spots matching across rounds
+    for r in sorted(set(round_dataframes.keys()) - {anchor_round, }):
+        query_df = round_dataframes[r]
+        query_coordinates = query_df[[Axes.ZPLANE, Axes.Y, Axes.X]]
+        nn = NearestNeighbors(n_neighbors=1)
```

Comment:
does NN not support a distance filter intrinsically?

Reply:
it looks like no there's a metric param that might work. But probs another question for @ambrosejcarr